### PR TITLE
metrics: standardize input naming + ordering

### LINKED ISSUE #262: Metrics: input naming and order conventions
While implementing the deterministic and probabilistic forecast metrics, I forgot to ensure that the function inputs following the same conventions regarding names and ordering of inputs. While the conventions within each type (deterministic and probabilistic) are consistent, they are not the same between the two:
- names:
   - deterministic: `y_true` (observations), `y_pred` (forecasts)
   - probabilistic: `obs` (observations), `fx` (forecasts)
- input order:
   - deterministic: observations, forecasts
   - probabilistic: forecasts, observations

For consistency, we should revise the metrics code to use a more consistent naming and ordering convention for the inputs.

### PR DESCRIPTION
**Pull request by dplarson**
- [x] Closes #262 .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Work on standardizing the naming and ordering conventions for inputs to all metric functions:
- naming: `obs` = observations, `fx` = forecasts
- order: `obs` then `fx` then other inputs (e.g. `obs, fx, norm`)

### CODE REVIEW

**Comment by dplarson** [comment]
@wholmgren Should I also add equations inside the docstrings for the deterministic metrics (to match the docstrings in the probabilistic metrics)? I can make a separate issue/PR if that's preferred, but it's easy enough to add them here.

**Comment by wholmgren** [comment]
@dplarson I think that's worthwhile.

**Comment by dplarson** [comment]
With 71da882 adding equations to the function docstrings, this PR is now ready for review.

File: solarforecastarbiter/metrics/deterministic.py [hunk 1]
```diff
@@ -19,218 +19,276 @@
...
-        True values.
-    y_pred : array-like
-        Predicted values.
-    y_ref: array_like
-        Reference forecast values.
+    obs : (n,) array-like
+        Observed values.
+    fx : (n,) array-like
+        Forecasted values.
+    ref : (n,) array_like
```
Comment (wholmgren) [inline solarforecastarbiter/metrics/deterministic.py]: maybe `fx_ref`? Just an idea.

File: solarforecastarbiter/metrics/deterministic.py [hunk 2]
```diff
@@ -19,218 +19,276 @@
...
+        A reference forecast.
```
Comment (wholmgren) [inline solarforecastarbiter/metrics/deterministic.py]: mild preference for the original wording. why the change?

File: solarforecastarbiter/metrics/deterministic.py [hunk 3]
```diff
@@ -19,218 +19,276 @@
...
-    return np.sum(x1 * x2) / (
-        np.sqrt(np.sum(x1 ** 2)) * np.sqrt(np.sum(x2 ** 2))
-    )
+    fx_avg = np.mean(fx)
+    obs_avg = np.mean(obs)
+    A = np.sum((fx - fx_avg) * (obs - obs_avg))
+    B = np.sqrt(np.sum((fx - fx_avg) ** 2))
+    C = np.sqrt(np.sum((obs - obs_avg) ** 2))
+    r = A / (B * C)
+    return r
```
Comment (wholmgren) [inline solarforecastarbiter/metrics/deterministic.py]: should have asked earlier... why don't we use [`numpy.corrcoef`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.corrcoef.html)?

File: solarforecastarbiter/metrics/probabilistic.py [hunk 4]
```diff
@@ -82,17 +85,15 @@ def brier_skill_score(fx, fx_prob, ref, ref_prob, obs):
         interval.
     ref_prob : (n,) array_like
```
Comment (wholmgren) [inline solarforecastarbiter/metrics/probabilistic.py]: I guess this would also become `fx_ref_prob` for consistency with `fx_ref` (*if* we decide to do that).

File: docs/source/whatsnew/1.0.0b3.rst [hunk 5]
```diff
@@ -22,6 +22,9 @@ API Changes
 * Add additional metric for probabilistic forecasts
   :py:mod:`solarforecastarbiter.metrics.probabilistic.continuous_ranked_probability_score`. (:issue:`250`) (:pull:`257`)
```
Comment (wholmgren) [inline docs/source/whatsnew/1.0.0b3.rst]: no empty line between bullets

File: solarforecastarbiter/metrics/deterministic.py (hunk 1 above)
Comment (dplarson) [reply solarforecastarbiter/metrics/deterministic.py]: Done.

File: solarforecastarbiter/metrics/deterministic.py (hunk 2 above)
Comment (dplarson) [reply solarforecastarbiter/metrics/deterministic.py]: To clarify: do you mean you prefer "Reference forecast values." instead of "A reference forecast" or are you referring to the previous "True values.", "Predicted values.", etc. in all the docstrings?

If the latter, my thought was that it's better to use "observation" with `obs` and "forecast" with `fx` to reinforce the variable naming choice. But I'm fine keeping the old wording if that's preferred. (I don't have a strong preference either way.)

File: solarforecastarbiter/metrics/deterministic.py (hunk 3 above)
Comment (dplarson) [reply solarforecastarbiter/metrics/deterministic.py]: Good idea. (I forgot about `numpy.corrcoef()`.)

I'll revise accordingly.

File: solarforecastarbiter/metrics/deterministic.py (hunk 2 above)
Comment (wholmgren) [reply solarforecastarbiter/metrics/deterministic.py]: > do you mean you prefer "Reference forecast values" instead of "A reference forecast"

This but not a big deal. Seems more consistent with the other doc strings

Comment (dplarson) [reply solarforecastarbiter/metrics/deterministic.py]: Gotcha. I'll revise to `Reference forecast values`. (I agree that sounds better and is more consistent.)

File: docs/source/whatsnew/1.0.0b3.rst (hunk 5 above)
Comment (dplarson) [reply docs/source/whatsnew/1.0.0b3.rst]: Corrected.

File: solarforecastarbiter/metrics/probabilistic.py (hunk 4 above)
Comment (dplarson) [reply solarforecastarbiter/metrics/probabilistic.py]: Yeah, that's the "trade-off"; `ref` and `ref_prob` or `fx_ref` and `fx_ref_prob`. I personally prefer `ref` and `ref_prob`, but I'm okay either way. So it's your call.

File: solarforecastarbiter/metrics/deterministic.py [hunk 6]
```diff
@@ -188,7 +189,7 @@ def pearson_correlation_coeff(obs, fx):
 
     """
 
-    r = np.corrcoef(obs, fx)[0, 1]
+    r, _ = sp.stats.pearsonr(obs, fx)
```
Comment (wholmgren) [inline solarforecastarbiter/metrics/deterministic.py]: out of curiosity, why scipy instead of numpy?

Comment (wholmgren) [reply solarforecastarbiter/metrics/deterministic.py]: nevermind, I see the detailed commit now

**Comment by dplarson** [comment]
The only "unsolved" issue is whether to use `ref` and `ref_prob` for the reference forecasts or `fx_ref` and `fx_ref_prob` (where `*_prob` are the probabilities in the probabilistic metrics). @wholmgren @alorenzo175 Any preference?

File: solarforecastarbiter/metrics/deterministic.py [hunk 7]
```diff
@@ -19,218 +20,271 @@
...
+        The NRMSE [%] of the forecast.
 
     """
 
-    return root_mean_square(y_true, y_pred) / y_norm * 100.0
+    return root_mean_square(obs, fx) / norm * 100.0
 
 
-def forecast_skill(y_true, y_pred, y_ref):
+def forecast_skill(obs, fx, fx_ref):
```
Comment (alorenzo175) [inline solarforecastarbiter/metrics/deterministic.py]: I guess I like ``ref`` slightly better than ``fx_ref``. In any case this should be ``ref`` for now to be consistent with the rest.

Comment (dplarson) [reply solarforecastarbiter/metrics/deterministic.py]: I'll standardize the code to use `ref` for reference forecasts.

**Comment by dplarson** [comment]
@wholmgren @alorenzo175 I've revised to use `ref` (and `ref_prob`) for the reference forecasts. With that, I think this PR can now be merged.

**Comment by wholmgren** [comment]
thanks @dplarson!
